src/sentry/incidents/logic.py

```diff
@@ -108,6 +115,110 @@ def create_incident(
     return incident
 
 
+INCIDENT_START_PERIOD = timedelta(days=14)
+INCIDENT_START_ROLLUP = timedelta(minutes=15)
+
+
+def calculate_incident_start(query, projects, groups):
+    """
+    Attempts to automatically calculate the date that an incident began at based
+    on the events related to the incident.
+    """
+    params = {}
+    if groups:
+        params['issue.id'] = [g.id for g in groups]
+        end = max(g.last_seen for g in groups) + timedelta(seconds=1)
+    else:
+        end = timezone.now()
+
+    params['start'] = end - INCIDENT_START_PERIOD
+    params['end'] = end
+
+    if projects:
+        params['project_id'] = [p.id for p in projects]
+
+    query_args = get_snuba_query_args(query, params)
+    rollup = int(INCIDENT_START_ROLLUP.total_seconds())
+
+    result = raw_query(
+        aggregations=[
+            ('count()', '', 'count'),
+            ('min', 'timestamp', 'first_seen'),
+        ],
+        orderby='time',
+        groupby=['time'],
+        rollup=rollup,
+        referrer='incidents.calculate_incident_start',
+        limit=10000,
+        **query_args
+    )['data']
+    # TODO: Start could be the period before the first period we find
+    result = zerofill(result, params['start'], params['end'], rollup, 'time')
+
+    # We want to linearly scale scores from 100% value at the most recent to
+    # 50% at the oldest. This gives a bias towards newer results.
+    negative_weight = (1.0 / len(result)) / 2
+    multiplier = 1.0
+    cur_spike_max_count = -1
+    cur_spike_start = None
+    cur_spike_end = None
+    max_height = 0
+    incident_start = None
+    cur_height = 0
+    prev_count = 0
+
+    def get_row_first_seen(row, default=None):
+        first_seen = default
+        if 'first_seen' in row:
+            first_seen = parse_date(row['first_seen']).replace(tzinfo=pytz.utc)
+        return first_seen
+
+    def calculate_start(spike_start, spike_end):
+        """
+        We arbitrarily choose a date about 1/3 into the incident period. We
+        could potentially improve this if we want by analyzing the period in
+        more detail and choosing a date that most closely fits with being 1/3
+        up the spike.
+        """
+        spike_length = (spike_end - spike_start)
+        return spike_start + (spike_length / 3)
+
+    for row in reversed(result):
+        cur_count = row.get('count', 0)
+        if cur_count < prev_count or cur_count > 0 and cur_count == prev_count:
+            cur_height = cur_spike_max_count - cur_count
+        elif cur_count > 0 or prev_count > 0 or cur_height > 0:
+            # Now we've got the height of the current spike, compare it to the
+            # current max. We decrease the value by `multiplier` so that we
+            # favour newer results
+            cur_height *= multiplier
+            if cur_height > max_height:
+                # If we detect that we have a new highest peak, then set a new
```

Comment:
It seems to me you are finding the incident start as the maximum variation between two consecutive data points (normalized through a multiplier). This does not take the amplitude of the spike into account. Is this intended? 
What if we have a relatively consistent issue whose number of event is mostly stable with minor variations. This would identify an incident start in that case as well no matter how small oscillations on the time series are.
Am I missing something ?

Reply:
This is only intended for detecting incident start dates on manually created incidents. So people choose some arbitrary set of groups, and then we try and determine when it started. Basically we don't want to force people to choose a start date themselves.

You're right that this doesn't work well on oscillating groups. It'll basically choose the first peak if they're close to the same size. Do you have any thoughts on how to improve?

I think we'll also have a different algorithm for determining the start date on rule based incidents, since I feel that in that case we'll want to choose the start date as part of the spike that occurred to push it over the threshold.

Reply:
"Do you have any thoughts on how to improve?"
It depends how far you want to go with accuracy. There are many algorithms to detect outliers and anomalies like STL: https://www.wessa.net/download/stl.pdf?source=post_page 
They can get quite complex, not sure how far you want to go with this feature.

Reply:
Honestly, not too far. I think that in terms of importance, manually created incidents aren't all that important once we have rule based creation and/or anomaly detection. My goal with this detection is to get a start date that looks decent. 

One concern I have is that with manually created incidents there may be no actual anomaly to even detect, so in some cases it's just going to not work all that great anyway.